### PR TITLE
#643 Group cli options in cli help

### PR DESCRIPTION
**Pull request by yashdsaraf**

### CODE REVIEW

File: src/scancode/cli.py [hunk 1]
```diff
@@ -231,11 +231,44 @@ class ScanCommand(BaseCommand):
...
+            ('scans', []),
+            ('output', []),
+            ('pre-scan', []),
+            ('post-scan', []),
+            ('misc', []),
+            ('core', []),
+        ])
+
+        for param in self.get_params(ctx):
+            rv = param.get_help_record(ctx)
```
Comment (pombredanne) [inline src/scancode/cli.py]: what is the meaning of `rv` here? May be a more explicit name would be better?

File: src/scancode/cli.py [hunk 2]
```diff
@@ -231,11 +231,44 @@ class ScanCommand(BaseCommand):
...
+                    opts['core'].append(rv)
+
+        with formatter.section('Options'):
+            for group, opt in opts.items():
+                if opt:
+                    with formatter.section(group):
+                        formatter.write_dl(opt)
+
+class ScanOption(click.Option):
+
```
Comment (pombredanne) [inline src/scancode/cli.py]: Can you document a bit what this is about?

File: src/scancode/cli.py [hunk 3]
```diff
@@ -231,11 +231,46 @@ class ScanCommand(BaseCommand):
     short_usage_help = '''
 Try 'scancode --help' for help on options and arguments.'''
 
-    def get_params(self, ctx):
-        """
-        Add options returned by plugins to the params list
-        """
-        return super(BaseCommand, self).get_params(ctx)
+    def format_options(self, ctx, formatter):
+        """Writes all the option groups into the formatter if they exist."""
```
Comment (pombredanne) [inline src/scancode/cli.py]: Can you use the convention we use else for docstrings ? e.g. on multiple lines and using imperative style (Write and not Writes)

File: src/scancode/cli.py [hunk 4]
```diff
@@ -231,11 +231,46 @@ class ScanCommand(BaseCommand):
...
+            ('scans', []),
+            ('output', []),
+            ('pre-scan', []),
+            ('post-scan', []),
+            ('misc', []),
+            ('core', []),
+        ])
+
+        for param in self.get_params(ctx):
+            help_record = param.get_help_record(ctx)
```
Comment (pombredanne) [inline src/scancode/cli.py]: What is help record?

File: src/scancode/cli.py [hunk 5]
```diff
@@ -231,11 +231,46 @@ class ScanCommand(BaseCommand):
...
+                    groups['misc'].append(help_record)
+
+        with formatter.section('Options'):
+            for group, option in groups.items():
+                if option:
+                    with formatter.section(group):
+                        formatter.write_dl(option)
+
+class ScanOption(click.Option):
+    """Allows an extra param `group` to be set which can be used
```
Comment (pombredanne) [inline src/scancode/cli.py]: Same comment as above about to docstring style.

File: tests/scancode/test_scan_help_groups.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    def test_scan_help_with_group(self):
+        @click.command(name='scan', cls=ScanCommand)
+        @click.option('--opt', is_flag=True, help='Help text for option', group='core', cls=ScanOption)
+        def scan(opt):
+            pass
+
+        runner = CliRunner()
+        result = runner.invoke(scan, ['--help'])
+        assert 'core:\n    --opt  Help text for option\n' in result.output
```
Comment (pombredanne) [inline tests/scancode/test_scan_help_groups.py]: Can you add a test to check that the current ScanCode cli help looks about the same as before?

File: src/scancode/cli.py (hunk 4 above)
Comment (yashdsaraf) [inline src/scancode/cli.py]: It's a list of the `option` name and its help text.

File: tests/scancode/test_scan_help_groups.py (hunk 6 above)
Comment (yashdsaraf) [inline tests/scancode/test_scan_help_groups.py]: I'm sorry I didn't really get that, what did you mean by `looks about the same as before`

File: src/scancode/cli.py [hunk 7]
```diff
@@ -231,11 +231,46 @@ class ScanCommand(BaseCommand):
...
+            ('output', []),
+            ('pre-scan', []),
+            ('post-scan', []),
+            ('misc', []),
+            ('core', []),
+        ])
+
+        for param in self.get_params(ctx):
+            help_record = param.get_help_record(ctx)
+            if help_record is not None:
```
Comment (pombredanne) [inline src/scancode/cli.py]: Do you really need to test for None instead of just `if help_record` ?

Comment (yashdsaraf) [inline src/scancode/cli.py]: That striked strange to me as well, but I picked up the code directly from click's core and modified it, hence I thought best not to mess with that `if`.

File: tests/scancode/test_scan_help_groups.py (hunk 6 above)
Comment (yashdsaraf) [inline tests/scancode/test_scan_help_groups.py]: @pombredanne ping?

Comment (pombredanne) [inline tests/scancode/test_scan_help_groups.py]: I meant by that that you should have a cli test that run `scancode -h` and check that the help returned looks like it should

Comment (yashdsaraf) [inline tests/scancode/test_scan_help_groups.py]: Oh, alright, I thought you meant something more complex than that :P

File: src/scancode/cli.py [hunk 8]
```diff
@@ -231,11 +231,51 @@ class ScanCommand(BaseCommand):
...
 Try 'scancode --help' for help on options and arguments.'''
 
-    def get_params(self, ctx):
+    def format_options(self, ctx, formatter):
         """
-        Add options returned by plugins to the params list
+        Write all options into the formatter in groups they belong to.
+        If a group is not defined, add the option to `misc` group
         """
-        return super(BaseCommand, self).get_params(ctx)
```
Comment (pombredanne) [inline src/scancode/cli.py]: Just so that I understand, you do not need to override `get_params` anymore?

File: src/scancode/cli.py (hunk 4 above)
Comment (pombredanne) [inline src/scancode/cli.py]: A comment telling this would be nice

File: src/scancode/cli.py [hunk 9]
```diff
@@ -231,11 +231,51 @@ class ScanCommand(BaseCommand):
...
+            ('pre-scan', []),
+            ('post-scan', []),
+            ('misc', []),
+            ('core', []),
+        ])
+
+        for param in self.get_params(ctx):
+            help_record = param.get_help_record(ctx)
+            if help_record:
+                if hasattr(param, 'group') and param.group:
```
Comment (pombredanne) [inline src/scancode/cli.py]: What about instead:
`param_group = getattr(param, 'group', None)`

File: src/scancode/cli.py [hunk 10]
```diff
@@ -231,11 +231,51 @@ class ScanCommand(BaseCommand):
...
+        groups = OrderedDict([
+            ('scans', []),
+            ('output', []),
+            ('pre-scan', []),
+            ('post-scan', []),
+            ('misc', []),
+            ('core', []),
+        ])
+
+        for param in self.get_params(ctx):
```
Comment (pombredanne) [inline src/scancode/cli.py]: is param an `Option` here?

File: src/scancode/cli.py [hunk 11]
```diff
@@ -231,11 +231,51 @@ class ScanCommand(BaseCommand):
     short_usage_help = '''
 Try 'scancode --help' for help on options and arguments.'''
 
-    def get_params(self, ctx):
+    def format_options(self, ctx, formatter):
```
Comment (pombredanne) [inline src/scancode/cli.py]: Just to make sure I get this: is this overriden from the click class?
If so, a mention in the docstring would be nice for future understaning

File: src/scancode/cli.py [hunk 12]
```diff
@@ -273,47 +313,47 @@ def validate_exclusive(ctx, exclusive_options):
...
-
-@click.help_option('-h', '--help')
-@click.option('--examples', is_flag=True, is_eager=True, callback=print_examples, help=('Show command examples and exit.'))
-@click.option('--about', is_flag=True, is_eager=True, callback=print_about, help='Show information about ScanCode and licensing and exit.')
-@click.option('--version', is_flag=True, is_eager=True, callback=print_version, help='Show the version and exit.')
-
-@click.option('--diag', is_flag=True, default=False, help='Include additional diagnostic information such as error messages or result details.')
-@click.option('--timeout', is_flag=False, default=DEFAULT_TIMEOUT, type=float, show_default=True, help='Stop scanning a file if scanning takes longer than a timeout in seconds.')
-@click.option('--reindex-licenses', is_flag=True, default=False, is_eager=True, callback=reindex_licenses, help='Force a check and possible reindexing of the cached license index.')
+              help=('Ignore files matching <pattern>.'), group='pre-scan', cls=ScanOption)
```
Comment (pombredanne) [inline src/scancode/cli.py]: Could you use a CONSTANT of sorts for the group values?

File: tests/scancode/data/help/help.txt [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  scan the <input> file or directory for origin clues and license and save
+  results to the <output_file>.
+
+  The scan results are printed to stdout if <output_file> is not provided.
+  Error and progress is printed to stderr.
+
+Options:
+
+  scans:
```
Comment (pombredanne) [inline tests/scancode/data/help/help.txt]: I am not 100% sure that we should include the "group" label in the output.... but this is good enough for now ;)

File: tests/scancode/test_scan_help_groups.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+    def test_scan_cli_help(self):
+        test_file = self.get_test_loc('help/help.txt')
```
Comment (pombredanne) [inline tests/scancode/test_scan_help_groups.py]: this is not a test file imho but an expected file instead

File: src/scancode/cli.py (hunk 8 above)
Comment (yashdsaraf) [inline src/scancode/cli.py]: No, that function is a leftover from when I was adding `add_cmdline_option`, anyway this method is properly overridden in #704

File: src/scancode/cli.py (hunk 10 above)
Comment (yashdsaraf) [inline src/scancode/cli.py]: Yes

File: src/scancode/cli.py (hunk 11 above)
Comment (yashdsaraf) [inline src/scancode/cli.py]: It is, I'll update it now

File: tests/scancode/data/help/help.txt (hunk 13 above)
Comment (yashdsaraf) [inline tests/scancode/data/help/help.txt]: This is how `pytest -h` looks like, but I can exclude the `group` labels and just group options using extra indentation, if that's required.

Comment (pombredanne) [inline tests/scancode/data/help/help.txt]: Thanks... keep it this way for now !

**Comment by pombredanne** [comment]
@yashdsaraf Looking all good to me ... Merging! Thanks!
